File: SelfCert/SelfCert/Connection.m
//
//  Connection.m
//  SelfCert
//
//  Created by Rob Napier on 11/25/12.
//  Copyright (c) 2012 Rob Napier. All rights reserved.
//

#import "Connection.h"

@interface Connection () <NSURLSessionDelegate>
@property (nonatomic, readwrite, strong) NSArray *anchors;
@property (nonatomic, readwrite, strong) NSURLSession *session;
@end

@implementation Connection

- (id)init {
    self = [super init];
    if (self) {
        NSError *error = NULL;
        NSString *path = [[NSBundle mainBundle] pathForResource:@"www.google.com"
                                                         ofType:@"cer"];
        NSData *certData = [NSData dataWithContentsOfFile:path
                                                  options:0
                                                    error:&error];

        if (! certData) {
            // Handle error reading
        }

        SecCertificateRef
        certificate = SecCertificateCreateWithData(NULL, (__bridge CFDataRef)certData);

        if (!certificate) {
            // Handle error parsing
        }

        self.anchors = [NSArray arrayWithObject:CFBridgingRelease(certificate)];

        self.session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:self delegateQueue:nil];
        NSURLSessionDataTask *task = [self.session dataTaskWithURL:[NSURL URLWithString:@"https://www.google.com"]
                                                 completionHandler: ^(NSData *data, NSURLResponse *response, NSError *error){
                                                     if (error != nil) {
                                                         NSLog(@"Failed: %@", error);
                                                     } else {
                                                         NSLog(@"Succeeded");
                                                     }
                                                 }];

        [task resume];
    }
    return self;
}

- (void)URLSession:(NSURLSession *)session didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential *credential))completionHandler {

    SecTrustRef trust = challenge.protectionSpace.serverTrust;

    SecTrustSetAnchorCertificates(trust, (__bridge CFArrayRef)self.anchors);
    SecTrustSetAnchorCertificatesOnly(trust, true);

    SecTrustResultType result;
    OSStatus status = SecTrustEvaluate(trust, &result);
    if (status == errSecSuccess &&
        (result == kSecTrustResultProceed ||
         result == kSecTrustResultUnspecified)) {
            
            NSURLCredential *cred = [NSURLCredential credentialForTrust:trust];
            completionHandler(NSURLSessionAuthChallengeUseCredential, cred);
        }
    else {
        completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, nil);
    }
}

@end

File: SelfCert/SelfCert/Connection.swift
//
//  Connection.swift
//  SelfCert
//
//  Created by Rob Napier on 3/27/16.
//  Copyright © 2016 Rob Napier. All rights reserved.
//

import Foundation

class Connection: NSObject, NSURLSessionDelegate {

    let url = NSURL(string: "https://www.google.com")!

    lazy private var anchors: [SecCertificate] = {
        let path = NSBundle.mainBundle().pathForResource("www.google.com", ofType: "cer")!
        let certData = NSData(contentsOfFile: path)!
        let certificate = SecCertificateCreateWithData(nil, certData)!
        return [certificate]
    }()

    lazy private var session: NSURLSession = {
        return NSURLSession(configuration: .defaultSessionConfiguration(),
                            delegate: self,
                            delegateQueue: nil)
    }()

    override init() {
        super.init()

        let task = session.dataTaskWithURL(url) { (data, response, error) in
            if let error = error {
                print("Failed: \(error)")
            } else {
                print("Succeeded")
            }
        }
        task.resume()
    }

    func URLSession(session: NSURLSession,
                    didReceiveChallenge challenge: NSURLAuthenticationChallenge,
                                        completionHandler: (NSURLSessionAuthChallengeDisposition, NSURLCredential?) -> Void) {

        let protectionSpace = challenge.protectionSpace

        if (protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust) {

            if let trust = protectionSpace.serverTrust {
                SecTrustSetAnchorCertificates(trust, anchors)
                SecTrustSetAnchorCertificatesOnly(trust, true)

                var result = SecTrustResultType(kSecTrustResultInvalid)
                let status = SecTrustEvaluate(trust, &result)

                if status == errSecSuccess {
                    switch Int(result) {
                    case kSecTrustResultProceed, kSecTrustResultUnspecified:
                        let cred = NSURLCredential(trust: trust)

                        completionHandler(.UseCredential, cred)
                        return

                    default:
                        print("Could not verify certificate: \(result)")
                    }
                }
            }

            // Something failed. Cancel
            completionHandler(.CancelAuthenticationChallenge, nil)

        } else {
            // We were asked for something other than server trust. Authenticate here if needed,
            // or call `completionHandler(.PerformDefaultHandling, nil)`
            preconditionFailure("Other authentication isn't implemented")
        }
    }
}